backfill: merge vectors from a temp forward index into a vector index

mw5h opened the pull request:
Previously mutation of values in a vector index during backfill was prohibited, effectively taking the table offline while the index is built. This patch does a few things:

* Re-enables mutations on vector indexes undergoing backfill.
* Changes the temporary index created for vector indexes to capture changes during backfill from a VECTOR index to a FORWARD index.
* Changes the schema of the temporary index so that it is keyed by the index prefix and primary key columns, with the vector as a payload column.
* Refactors the mvcc merger to take a functor argument to merge individual entries, with the existing kv merger as the default.
* Adds a second merger that re-encodes vector index entries read from the temporary index.

Because forward indexes don't record values for tombstone entries, we don't delete vectors from the vector index that are deleted during backfill. Since reading from a vector index requires an index join on the PK, this doesn't create a correctness issue, but it could cause accuracy issues if a large number of vector index entries are deleted during backfill.

Fixes: #144443
Release note (sql change): Tables with vector indexes will no longer be taken offline while the vector index builds.

cockroach-teamcity commented:
<details><summary><strong>⚪ Sysbench</strong> [SQL, 3node, oltp_read_write]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 7.881m ±0% | 7.894m ±0% | ~ | p=0.267 n=15    |
| ⚪ **allocs/op** | 7.817k ±2% | 7.831k ±2% | ~ | p=0.935 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/bae5ac8/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/bae5ac8ea5558f68cc9431b0b99bcf329f0f8db3/bin/pkg_sql_tests benchdiff/bae5ac8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/bae5ac8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/50fb4c1/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/50fb4c17afb98cd2e6ee3837c642e94772e84bc4/bin/pkg_sql_tests benchdiff/50fb4c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/50fb4c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/SQL/3node/oltp_read_write$ --old=50fb4c1 --new=bae5ac8 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>🔴 Sysbench</strong> [KV, 3node, oltp_read_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| 🔴 **sec/op** | 1.059m ±1% | 1.071m ±1% | +1.14% | p=0.000 n=15    |
| ⚪ **allocs/op** | 1.861k ±0% | 1.864k ±0% | ~ | p=0.114 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/bae5ac8/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/bae5ac8ea5558f68cc9431b0b99bcf329f0f8db3/bin/pkg_sql_tests benchdiff/bae5ac8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/bae5ac8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/50fb4c1/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/50fb4c17afb98cd2e6ee3837c642e94772e84bc4/bin/pkg_sql_tests benchdiff/50fb4c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/50fb4c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_read_only$ --old=50fb4c1 --new=bae5ac8 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_write_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 1.878m ±1% | 1.882m ±1% | ~ | p=0.081 n=15    |
| ⚪ **allocs/op** | 3.781k ±0% | 3.783k ±0% | ~ | p=0.753 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/bae5ac8/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/bae5ac8ea5558f68cc9431b0b99bcf329f0f8db3/bin/pkg_sql_tests benchdiff/bae5ac8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/bae5ac8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/50fb4c1/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/50fb4c17afb98cd2e6ee3837c642e94772e84bc4/bin/pkg_sql_tests benchdiff/50fb4c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/50fb4c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_write_only$ --old=50fb4c1 --new=bae5ac8 ./pkg/sql/tests
```

</details>

</details>

<details><summary>Artifacts</summary>

**download**:
```shell
mkdir -p new
gcloud storage cp gs://cockroach-microbench-ci/artifacts/bae5ac8ea5558f68cc9431b0b99bcf329f0f8db3/16730915743-1/\* new/
mkdir -p old
gcloud storage cp gs://cockroach-microbench-ci/artifacts/50fb4c17afb98cd2e6ee3837c642e94772e84bc4/16730915743-1/\* old/
```

</details>

_built with commit: [bae5ac8ea5558f68cc9431b0b99bcf329f0f8db3](https://github.com/cockroachdb/cockroach/commit/bae5ac8ea5558f68cc9431b0b99bcf329f0f8db3)_

<!-- microbench-ci -->

cockroach-teamcity commented:
<details><summary><strong>⚪ Sysbench</strong> [SQL, 3node, oltp_read_write]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 7.885m ±0% | 7.884m ±0% | ~ | p=0.567 n=15    |
| ⚪ **allocs/op** | 7.829k ±2% | 7.854k ±1% | ~ | p=0.894 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/8b69700/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/8b69700ed30abbfd1b8903ed52223d5c87dc0c96/bin/pkg_sql_tests benchdiff/8b69700/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/8b69700/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/50fb4c1/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/50fb4c17afb98cd2e6ee3837c642e94772e84bc4/bin/pkg_sql_tests benchdiff/50fb4c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/50fb4c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/SQL/3node/oltp_read_write$ --old=50fb4c1 --new=8b69700 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_read_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 1.084m ±1% | 1.094m ±1% | ~ | p=0.050 n=15    |
| ⚪ **allocs/op** | 1.864k ±0% | 1.864k ±0% | ~ | p=0.164 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/8b69700/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/8b69700ed30abbfd1b8903ed52223d5c87dc0c96/bin/pkg_sql_tests benchdiff/8b69700/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/8b69700/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/50fb4c1/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/50fb4c17afb98cd2e6ee3837c642e94772e84bc4/bin/pkg_sql_tests benchdiff/50fb4c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/50fb4c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_read_only$ --old=50fb4c1 --new=8b69700 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_write_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 1.889m ±0% | 1.889m ±0% | ~ | p=0.624 n=15    |
| ⚪ **allocs/op** | 3.784k ±0% | 3.787k ±0% | ~ | p=0.149 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/8b69700/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/8b69700ed30abbfd1b8903ed52223d5c87dc0c96/bin/pkg_sql_tests benchdiff/8b69700/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/8b69700/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/50fb4c1/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/50fb4c17afb98cd2e6ee3837c642e94772e84bc4/bin/pkg_sql_tests benchdiff/50fb4c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/50fb4c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_write_only$ --old=50fb4c1 --new=8b69700 ./pkg/sql/tests
```

</details>

</details>

<details><summary>Artifacts</summary>

**download**:
```shell
mkdir -p new
gcloud storage cp gs://cockroach-microbench-ci/artifacts/8b69700ed30abbfd1b8903ed52223d5c87dc0c96/16763310268-1/\* new/
mkdir -p old
gcloud storage cp gs://cockroach-microbench-ci/artifacts/50fb4c17afb98cd2e6ee3837c642e94772e84bc4/16763310268-1/\* old/
```

</details>

_built with commit: [8b69700ed30abbfd1b8903ed52223d5c87dc0c96](https://github.com/cockroachdb/cockroach/commit/8b69700ed30abbfd1b8903ed52223d5c87dc0c96)_

<!-- microbench-ci -->

DrewKimball reviewed (commented):
> Reviewed 6 of 15 files at r1, 3 of 7 files at r2, 2 of 2 files at r3, 1 of 2 files at r4, 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151074)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball)

___
*[`pkg/sql/backfill/backfill.go` line 609 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OX5PYovAId4Qgo_PWKj:-OX5PYovAId4Qgo_PWKk:b-fxds06) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e4df8a5fe247fd093e67c1dedf3e14ed8d035be/pkg/sql/backfill/backfill.go#L609)):*
> ```Go
> // contains the state that is stable over the lifetime of the merge operation
> // and is safe to share across different go routines.
> type VectorIndexMergeHelper struct {
> ```

nit: Could we pull out a common `vectorIndexHelperBase` struct?
___
*[`pkg/sql/backfill/mvcc_index_merger.go` line 158 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OXAeopB63iNdlCCaoPE:-OXAeopB63iNdlCCaoPF:b-gn5idb) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e4df8a5fe247fd093e67c1dedf3e14ed8d035be/pkg/sql/backfill/mvcc_index_merger.go#L158)):*
> ```Go
> 
> 		var vim VectorIndexMergeHelper
> 		if err := vim.Init(ctx, ibm.flowCtx.EvalCtx, ibm.desc, ibm.flowCtx.Cfg.VecIndexManager.(*vecindex.Manager), idxDesc, idxDesc); err != nil {
> ```

`...idxDesc, idxDesc); err != nil {` is this right?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 768 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OXAiyae25qnYZ7yeDUr:-OXAiyae25qnYZ7yeDUs:bxej71t) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e4df8a5fe247fd093e67c1dedf3e14ed8d035be/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L768)):*
> ```Go
> // fixupPartitioningForTempVectorIndex updates the partitioning for a temporary
> // vector index to inherit from the primary key instead of the vector index.
> func fixupPartitioningForTempVectorIndex(b BuildCtx, tempIdxSpec *indexSpec, tempID catid.IndexID) {
> ```

Do we have a test for this?
___
*[`pkg/sql/backfill/backfill_test.go` line 150 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OXAhZHD-t6jASdcYUhZ:-OXAhZHD-t6jASdcYUh_:bq2jgu5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e4df8a5fe247fd093e67c1dedf3e14ed8d035be/pkg/sql/backfill/backfill_test.go#L150)):*
> ```Go
> 		_, createIndexErr = db.ExecContext(ctx, `CREATE VECTOR INDEX vec_idx ON vectors (vec)`)
> 		if createIndexErr != nil {
> 			backfillBlocked <- struct{}{}
> ```

Should we just close the channel?
___
*[`pkg/sql/backfill/backfill_test.go` line 335 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OXAiiFHBdgs_w5puPtK:-OXAiiFHBdgs_w5puPtL:b-pjrdib) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e4df8a5fe247fd093e67c1dedf3e14ed8d035be/pkg/sql/backfill/backfill_test.go#L335)):*
> ```Go
> 
> 	// Search individually for vectors from the second set of inserts (1001-1100)
> 	// We should find at least 90% of those vectors
> ```

Is the test deterministic enough to do a more specific check? If not, do you know why?


<!-- Sent from Reviewable.io -->

mw5h reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151074)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @DrewKimball)

___
*[`pkg/sql/backfill/mvcc_index_merger.go` line 158 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OXAeopB63iNdlCCaoPE:-OXVLIhQ6jqAw8MUDYwg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e4df8a5fe247fd093e67c1dedf3e14ed8d035be/pkg/sql/backfill/mvcc_index_merger.go#L158)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

`...idxDesc, idxDesc); err != nil {` is this right?
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 768 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OXAiyae25qnYZ7yeDUr:-OXVLA-n5MojWYYii2Qt:b-rqma0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e4df8a5fe247fd093e67c1dedf3e14ed8d035be/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L768)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Do we have a test for this?
</blockquote></details>

pkg/ccl/logictestccl/testdata/logic_test/partitioning_index is what discovered the need for this in the first place.
___
*[`pkg/sql/backfill/backfill_test.go` line 150 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OXAhZHD-t6jASdcYUhZ:-OXVLEyU0wSOsYpbIMzR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e4df8a5fe247fd093e67c1dedf3e14ed8d035be/pkg/sql/backfill/backfill_test.go#L150)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Should we just close the channel?
</blockquote></details>

Done.
___
*[`pkg/sql/backfill/backfill_test.go` line 335 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OXAiiFHBdgs_w5puPtK:-OXVLFil3qOsmy000yxv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e4df8a5fe247fd093e67c1dedf3e14ed8d035be/pkg/sql/backfill/backfill_test.go#L335)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Is the test deterministic enough to do a more specific check? If not, do you know why?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

fqazi reviewed (changes requested):
Can you add a DML injection test in the schema changer framework, you can use the following as templates: pkg/sql/schemachanger/testdata/end_to_end/create_index/create_index.definition. The second thing worth updating tests that recreate secondary indexes due to alter primary key:   pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_primary_key_vanilla/alter_table_alter_primary_key_vanilla.definition.

Theses tests allow us to inject IUD operations during each phase of the schema change as a sanity check

> @fqazi reviewed 6 of 15 files at r1, 2 of 7 files at r2, 2 of 2 files at r3, 2 of 5 files at r5, 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-:-OYC9ua65n-ZA2ofvdPj:bcjntlr)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @DrewKimball, and @spilchen)

___
*[`pkg/sql/backfill/backfill.go` line 692 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OYC7osw4PfHxZgRMYIu:-OYC7osw4PfHxZgRMYIv:b-gesbxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f25c612940727bf16a752761e906a89b2ae345d/pkg/sql/backfill/backfill.go#L692)):*
> ```Go
> ) error {
> 	// Initialize the base fields using the shared initialization
> 	if err := vimh.initBase(ctx, evalCtx, desc, destIndex, desc.GetPrimaryIndex(), vecIndexManager); err != nil {
> ```

One thing to confirm is if the right thing happens if we recreate an index after a primary key swap
___
*[`pkg/sql/backfill/backfill.go` line 716 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OYC8w7TAwjUSkbgzEgo:-OYC8w7TAwjUSkbgzEgp:bad6nij) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f25c612940727bf16a752761e906a89b2ae345d/pkg/sql/backfill/backfill.go#L716)):*
> ```Go
> 		// Avoid duplicates if PK column is already in the vector index
> 		found := false
> 		for _, existingColID := range columnsToFetch {
> ```

Might be cleaner to just keep a Set and look up against that.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> @DrewKimball reviewed 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151074)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @mw5h, and @spilchen)

___
*[`pkg/sql/backfill/backfill.go` line 658 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OYIJuvr7AFYpODoG05l:-OYIJuvr7AFYpODoG05m:btm0ak5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f25c612940727bf16a752761e906a89b2ae345d/pkg/sql/backfill/backfill.go#L658)):*
> ```Go
> 	primaryIndex catalog.Index
> 	// builder is the span builder for reading from the primary key.
> 	builder span.Builder
> ```

I'm not sure if `span.Builder` is thread-safe, since it uses an internal `tree.DatumAlloc`. Maybe it should live on `VectorIndexMergerTxn`? Or maybe we could pull out a struct that only contains the thread-safe part of `span.Builder`
___
*[`pkg/sql/backfill/backfill.go` line 819 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OYIUchTAeicJMu8MeH1:-OYIUchTAeicJMu8MeH2:b-twoklz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f25c612940727bf16a752761e906a89b2ae345d/pkg/sql/backfill/backfill.go#L819)):*
> ```Go
> 	if tempWrapper.Deleted {
> 		// Not really anything we can do about a deleted key since that means
> 		// the vector in the PK is already deleted.
> ```

I guess in theory we could do a stale read...
___
*[`pkg/sql/backfill/backfill.go` line 828 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OYIUVCa3idOiTxAjMZZ:-OYIUVCb8g3FxbmLDgHU:bqieozz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f25c612940727bf16a752761e906a89b2ae345d/pkg/sql/backfill/backfill.go#L828)):*
> ```Go
> 
> 	// Decode the key bytes into EncDatumRow so we can use the span.Builder
> 	// properly. Use pre-allocated memory from MergeContext.
> ```

nit: `MergeContext` looks like it's a stale comment?
___
*[`pkg/sql/backfill/backfill.go` line 877 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OYIV3jd124qqismsO3y:-OYIV3jd124qqismsO3z:beworbv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f25c612940727bf16a752761e906a89b2ae345d/pkg/sql/backfill/backfill.go#L877)):*
> ```Go
> 	vectorDatum := vm.rowDatums[vm.vim.vectorPos]
> 	if vectorDatum == tree.DNull {
> 		// Null vector, skip this entry
> ```

I wonder if we could make the temp index a partial index that ignores null vectors.
___
*[`pkg/sql/backfill/backfill.go` line 897 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OYIWjOy0_AeDHyTSEDS:-OYIWjOy0_AeDHyTSEDT:b6cnw0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f25c612940727bf16a752761e906a89b2ae345d/pkg/sql/backfill/backfill.go#L897)):*
> ```Go
> 		prefixColDatum := vm.rowDatums[prefixColPos]
> 		encDatum := rowenc.DatumToEncDatum(prefixColType, prefixColDatum)
> 		prefixKey, err = encDatum.Encode(
> ```

Can we use `rowenc.EncodePartialIndexKey` here?


<!-- Sent from Reviewable.io -->

mw5h commented:
The DML injection tests are currently failing because I haven't figured out the right way to turn on the vector index feature flag.

Other comments I believe are all addressed.

mw5h reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151074)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @DrewKimball, @fqazi, and @spilchen)

___
*[`pkg/sql/backfill/backfill.go` line 658 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OYIJuvr7AFYpODoG05l:-OYiA4SA257EuCxUyimS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f25c612940727bf16a752761e906a89b2ae345d/pkg/sql/backfill/backfill.go#L658)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I'm not sure if `span.Builder` is thread-safe, since it uses an internal `tree.DatumAlloc`. Maybe it should live on `VectorIndexMergerTxn`? Or maybe we could pull out a struct that only contains the thread-safe part of `span.Builder`
</blockquote></details>

Done.
___
*[`pkg/sql/backfill/backfill.go` line 716 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OYC8w7TAwjUSkbgzEgo:-OYiA7Df4AAFEOdML3Z2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f25c612940727bf16a752761e906a89b2ae345d/pkg/sql/backfill/backfill.go#L716)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Might be cleaner to just keep a Set and look up against that.
</blockquote></details>

Done.
___
*[`pkg/sql/backfill/backfill.go` line 877 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OYIV3jd124qqismsO3y:-OYiAJ1H4q_MfBLWPQA1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f25c612940727bf16a752761e906a89b2ae345d/pkg/sql/backfill/backfill.go#L877)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I wonder if we could make the temp index a partial index that ignores null vectors.
</blockquote></details>

Done.
___
*[`pkg/sql/backfill/backfill.go` line 897 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OYIWjOy0_AeDHyTSEDS:-OYiAJhs5MCXb7-keOlN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f25c612940727bf16a752761e906a89b2ae345d/pkg/sql/backfill/backfill.go#L897)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Can we use `rowenc.EncodePartialIndexKey` here?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
@mw5h Is the `SET CLUSTER SETTING` not taking effect leading to the DML injection failures? Another option is we could just force it enabled for this test suite inside: pkg/ccl/schemachangerccl/multiregion_testcluster_factory.go and pkg/sql/schemachanger/sctest/test_server_factory.go

> @fqazi reviewed 6 of 6 files at r7, 30 of 30 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-:-OYl8gqtE4kZLIW424lx:byyxfa8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @DrewKimball, @mw5h, and @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 824 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OYl72cXD8KAU357sTPj:-OYl72cXD8KAU357sTPk:b-k53crg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3f21a3ad4cd0274a634b4cc15da5160a3fcac1d/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L824)):*
> ```Go
> 	// changed to add the vector column.
> 	// TODO(mw5h) In this case, will every value inserted into the temporary
> 	// index be NULL? If so, let's just not add a temporary index.
> ```

Yeah it would imply every value is non-null.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
@mw5h Also nice work with the schema changer tests!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-:-OYlBupV1KhJ8BSdPfSg:bjzui4x)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @DrewKimball, @mw5h, and @spilchen)



<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151074)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @DrewKimball, @mw5h, and @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 824 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OYl72cXD8KAU357sTPj:-OYlf8Or8x-a-Xw5_noZ:b-y5wap0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3f21a3ad4cd0274a634b4cc15da5160a3fcac1d/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L824)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

NULL or NOT NULL? If the column isn't in the current PK, will the temporary index ever see a non-NULL value for that column? If the answer is no, we can skip creating and writing to the temporary index altogether, which is a small performance win.
</blockquote></details>

Sorry yes all values will be null since the column is being removed.


<!-- Sent from Reviewable.io -->

mw5h committed:
sctest: pass around cluster settings instead of always generating new

Previously, the end_to_end test would always create a new set of cluster
settings when spawning a subtest, causing cluster settings made during
setup to be lost. This change causes subtests to inherit settings,
retaining them.

Release note: None

mw5h committed:
vecstore: search all PKs for get full vector index

When we're changing primary keys, it's possible that the PrimaryKey()
isn't actually keyed by the keys embedded as suffixes in the vector
index. Since we need these indexes to be able to execute
vecstore.GetFullVectors(), build the GetFullVectorsFetchSpec using the
first primary key encoded index that has the keys that we have embedded
in our key.

This change addresses a bug discovered during construction of the schema
changer DML injection tests.

Release note: None

DrewKimball reviewed (approved):
Great work! Everything <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> apart from a question about the schema-changer tests.

> @DrewKimball reviewed 5 of 6 files at r7, 5 of 50 files at r9, 16 of 16 files at r11, 32 of 32 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-:-OZITiNT3gh8a6TmHLl5:b-dr38a3)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @mw5h, and @spilchen)

___
*[`-- commits` line 19 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OZIQAs_C5N67eMLxTkM:-OZIQAs_C5N67eMLxTkN:bfmoc7r):*
Hmm, so the issue here is a new vector index referring to the old primary key and/or an old vector index referring to the new primary key?
___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_primary_key_vector/alter_table_alter_primary_key_vector.definition` line 24 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OZIT9z30uXcRpG4jV3K:-OZIT9z30uXcRpG4jV3L:b-fzgn1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c58c689311810d4c24dbb2b905513a76ae8d792/pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_primary_key_vector/alter_table_alter_primary_key_vector.definition#L24)):*
> ```
> # Verify vector data is correctly indexed during recreation
> stage-query phase=PostCommitPhase stage=:
> SELECT count(*) FROM t WHERE embedding IS NOT NULL;
> ```

This doesn't actually use the index, does it?
___
*[`pkg/sql/backfill/backfill.go` line 705 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OZIQu-d6mqcP2MTf5bZ:-OZIQu-d6mqcP2MTf5b_:b-83sjo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c58c689311810d4c24dbb2b905513a76ae8d792/pkg/sql/backfill/backfill.go#L705)):*
> ```Go
> 
> 	// Add all the vector index columns and then any unused PK columns
> 	for _, index := range []catalog.Index{destIndex, vimh.primaryIndex} {
> ```

Nice.


<!-- Sent from Reviewable.io -->

mw5h committed:
vecstore: encoding cleanup

No functional change here, just some code cleanup.

Epic: None

mw5h committed:
backfill: merge vectors from a temp forward index into a vector index

Previously mutation of values in a vector index during backfill was
prohibited, effectively taking the table offline while the index is
built. This patch does a few things:

* Re-enables mutations on vector indexes undergoing backfill.
* Changes the temporary index created for vector indexes to capture
  changes during backfill from a VECTOR index to a FORWARD index.
* Changes the schema of the temporary index so that it is keyed by the
  index prefix and primary key columns, with the vector as a payload
  column.
* Refactors the mvcc merger to take a functor argument to merge
  individual entries, with the existing kv merger as the default.
* Adds a second merger that re-encodes vector index entries read from
  the temporary index.

Because forward indexes don't record values for tombstone entries, we
don't delete vectors from the vector index that are deleted during
backfill. Since reading from a vector index requires an index join on
the PK, this doesn't create a correctness issue, but it could cause
accuracy issues if a large number of vector index entries are deleted
during backfill.

Fixes: #144443
Release note (sql change): Tables with vector indexes will no longer be
taken offline while the vector index builds.

mw5h commented:
I've cleaned up the patches a little bit and have removed the schemachanger DML injection tests for now. Those tests won't pass until [#153684](https://github.com/cockroachdb/cockroach/pull/153684) or something similar goes in (the tests I have do pass with that patch applied). Since we're getting close to branch cut and this is a moderately beefy patch, I'd like to commit this now and then bring the extra tests in once @fqazi's patch goes in.

@fqazi Do you think you could give this a review for Foundations? I spoke to @spilchen previously and he seemed to be fine with you continuing the review.

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @fqazi reviewed 4 of 50 files at r9, 48 of 48 files at r13, 1 of 1 files at r14, 50 of 50 files at r16, 3 of 3 files at r17, 47 of 47 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-:-O_Svv2k3M8WCUmxeQNb:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @DrewKimball, @mw5h, and @spilchen)

___
*[`pkg/sql/vecindex/vecstore/store_txn.go` line 377 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-O_Sv-c9CY34qifu3ylv:-O_Sv-c9CY34qifu3ylw:b3she9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c40c63685fefece7430b515dcdf402be92b1481a/pkg/sql/vecindex/vecstore/store_txn.go#L377)):*
> ```Go
> 		}
> 
> 		primaryKey = idx
> ```

Might be good to assert this is the highest index ID.


<!-- Sent from Reviewable.io -->

mw5h reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151074)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @DrewKimball, @fqazi, and @spilchen)

___
*[`-- commits` line 19 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OZIQAs_C5N67eMLxTkM:-OZR5TpE9AQ-pgTlXVe2:b-u536iq):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Hmm, so the issue here is a new vector index referring to the old primary key and/or an old vector index referring to the new primary key?
</blockquote></details>

The vector index is nominally not changing, but has to be rebuilt because the PK is changing. The relation will have a new secondary index with primary index encoding (the new primary being built). There will also be two vector indexes, one referring to each primary. Neither of these vector indexes will be marked as mutations because they're "not changing". Because of how vector indexes work, however, both of these indexes need to work, so we have to wire them up.
___
*[`pkg/sql/backfill/backfill.go` line 692 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OYC7osw4PfHxZgRMYIu:-O_SyYXo1rOaBZbIMM0C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f25c612940727bf16a752761e906a89b2ae345d/pkg/sql/backfill/backfill.go#L692)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

One thing to confirm is if the right thing happens if we recreate an index after a primary key swap
</blockquote></details>

Done.
___
*[`pkg/sql/backfill/backfill.go` line 819 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OYIUchTAeicJMu8MeH1:-OZR6LW7A2dDEWE77tGj:b-qqjjc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f25c612940727bf16a752761e906a89b2ae345d/pkg/sql/backfill/backfill.go#L819)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I guess in theory we could do a stale read...
</blockquote></details>

Maybe not right now, if that's okay. We have the new backfill in the pipe and I'd rather chase that than chase corner cases.
___
*[`pkg/sql/backfill/backfill.go` line 877 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OYIV3jd124qqismsO3y:-OZR9xP93P_3Shm0WO3P:b-k93kz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f25c612940727bf16a752761e906a89b2ae345d/pkg/sql/backfill/backfill.go#L877)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Turns out that this doesn't work if the vector is an expression because the expression column can't be found when we're type checking.
___
*[`pkg/sql/backfill/backfill.go` line 705 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OZIQu-d6mqcP2MTf5bZ:-OZR6b4R7_Xm2uXOADxz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c58c689311810d4c24dbb2b905513a76ae8d792/pkg/sql/backfill/backfill.go#L705)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Nice.
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/vecstore/store_txn.go` line 377 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-O_Sv-c9CY34qifu3ylv:-O_SyOhT36R8D0gPq4_l:b-a6957n) ([raw file](https://github.com/cockroachdb/cockroach/blob/c40c63685fefece7430b515dcdf402be92b1481a/pkg/sql/vecindex/vecstore/store_txn.go#L377)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Might be good to assert this is the highest index ID.
</blockquote></details>

As discussed in slack, we think this logic works. 
___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_primary_key_vector/alter_table_alter_primary_key_vector.definition` line 24 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/151074#-OZIT9z30uXcRpG4jV3K:-OZR6ft43kF9I_c92dw-:brmlcsf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c58c689311810d4c24dbb2b905513a76ae8d792/pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_primary_key_vector/alter_table_alter_primary_key_vector.definition#L24)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

This doesn't actually use the index, does it?
</blockquote></details>

Oh good point.


<!-- Sent from Reviewable.io -->
